kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/source/reader/ReaderManagerService.scala

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017-2023 Lenses.io Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.lenses.streamreactor.connect.aws.s3.source.reader
+
+import cats.effect.IO
+import cats.effect.unsafe.implicits.global
+import com.typesafe.scalalogging.LazyLogging
+import io.lenses.streamreactor.connect.aws.s3.model.location.RemoteS3RootLocation
+import io.lenses.streamreactor.connect.aws.s3.source.config.PartitionSearcherOptions
+import io.lenses.streamreactor.connect.aws.s3.source.distribution.PartitionSearcher
+import io.lenses.streamreactor.connect.aws.s3.source.distribution.PartitionSearcherResponse
+
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicReference
+
+case class ReaderManagerState(
+  partitionResponses: Seq[PartitionSearcherResponse],
+  readerManagers:     Seq[ReaderManager],
+) {
+  def lastSearchTime: Option[Instant] = partitionResponses.map(_.lastSearchTime).minOption
+
+}
+
+class ReaderManagerService(
+  settings:              PartitionSearcherOptions,
+  partitionSearcher:     PartitionSearcher,
+  readerManagerCreateFn: (RemoteS3RootLocation, String) => ReaderManager,
+) extends LazyLogging {
+
+  private val readerManagerState: AtomicReference[ReaderManagerState] =
+    new AtomicReference(ReaderManagerState(Seq.empty, Seq.empty))
+
+  def getReaderManagers: Seq[ReaderManager] = {
+    launchDiscover().unsafeRunSync()
+    readerManagerState.get().readerManagers
+  }
+
+  private def launchDiscover(): IO[Unit] = {
+
+    val lastSearchTime = readerManagerState.get().lastSearchTime
+    val rediscoverDue  = settings.rediscoverDue(lastSearchTime)
+
+    if (rediscoverDue && settings.blockOnSearch) {
+      rediscover()
+    } else if (rediscoverDue && !settings.blockOnSearch) {
+      rediscover().start *> IO.unit
+    } else {
+      IO.unit
+    }
+  }
+
+  private def rediscover(): IO[Unit] =
```

Comment:
using a Ref, and considering there is one writer only I would suggest something like:

```
IO[ReaderManagerState] =
    for {
      _        <- IO(logger.debug("calculating updated state"))
      ioState  <- ref.get
      newParts <- partitionSearcher.findNewPartitions(ioState.partitionResponses)
      newReaderManagers = newParts
        .flatMap(part => part.results.partitions.map(part.root -> _))
        .map {
          case (location, res) => readerManagerCreateFn(location, res)
        }
        newState = ioState.copy(
      partitionResponses = newParts,
      readerManagers     = ioState.readerManagers ++ newReaderManagers,
    )
     _ <- refState.set(newState)
    } yield ()
```

I am not sure if it compiles. Happy to go on a call. 

This solution avoids creating an IO to get the value of the AtomicReference and invoke unsafeRunSync

Reply:
Yeah, I also agree with the usage of Ref here, I already mentioned above in a different file

Reply:
Fixed!